compute/etc/sql_exporter/retained_wal.sql

```diff
@@ -1,5 +1,10 @@
 SELECT
   slot_name,
-  pg_wal_lsn_diff(pg_current_wal_lsn(), restart_lsn)::FLOAT8 AS retained_wal
+  pg_wal_lsn_diff(
+    CASE
+      WHEN pg_is_in_recovery() THEN pg_last_wal_replay_lsn()
+      ELSE pg_current_wal_lsn()
+    END,
+    restart_lsn)::FLOAT8 AS retained_wal
 FROM pg_replication_slots
```

Comment:
I think there should be zero slots on replica, so we should never hit this, isn't it?

Reply:
Here is some output from the current support escalation:

```
postgres=# SELECT * FROM pg_replication_slots;
       slot_name       |  plugin  | slot_type | datoid | database | temporary | active | active_pid | xmin | catalog_xmin | restart_lsn  | confirmed_flush_lsn | wal_status | safe_wal_size | two_phase
-----------------------+----------+-----------+--------+----------+-----------+--------+------------+------+--------------+--------------+---------------------+------------+---------------+-----------
 prisma_pulse_c5b3649d | pgoutput | logical   |  32785 | redacted  | f         | f      |            |      |     59027802 | 363/B0B2D7C0 | 363/B0B2D7C1        | reserved   |               | f
 prisma_
(2 rows)
```

Reply:
I don't know what this prisma thing is.

Reply:
https://www.prisma.io/data-platform/pulse

Reply:
I see, but the fix still looks a bit strange to me. Slot only retains WAL at primary, here we will report a diff between pg_last_wal_replay_lsn() and slot restart_lsn, but it doesn't make much sense. Or does it?

Will it be better to just return nothing (or zero) in case of pg_is_in_recovery()?

Reply:
Yes. Just created test for it - combination of logical and physical replication.
If subscription is created prior to start replica, than slot is included in basebackup and so "inherited" by replica.
As far as this lost is never advanced, it efficiently blocks WAL truncation at replica.  
This is result:
```
kk@Konstantins-MacBook-Pro endpoints % du -sh *
 65M	primary
389M	secondary
```
So we should avoid copying  replication slots to replica. At least now, until wee support cascading replication.
It can be done in two ways:
1. Drop logical replication slots while starting standby.
2. Avoid inclusion of slots AUX files in basebackup for replica.
2) seems to be much natural and easier solution. The only problem is that PS should be somehow notified that It is starting replica. I don't know - may be it is already possible to determine (from some configuration parameters). In the worst case it will require some changes on compute_ctl and control plane.

Any other idea?

Reply:
I think 2. would require passing some argument to basebackup, so it should be harder to do compared to 1., which I think is just removing some files after unarchiving the basebackup, right?

Reply:
Yes, we can drop or disable slots. May be it can be done in Neon extension, so do not require changing Postgres core. But I am not sure how if it is possible (is it possible to drop slot at replica?).  Certainly deleting files is always possible, but there is race condition - Postgres may already created slots based this files.
This is why I think 2) can be safer and straightforward solution.

Reply:
>  Certainly deleting files is always possible, but there is race condition - Postgres may already created slots based this files.

We first do basebackup, then start Postgres; so I think if we drop files before starting Postgres, then there is no race, or is it?

I only wonder about further streaming. Yanic said that once you create a slot at primary it's immediately availabe at replica. So looks like it's streamed? If so, then just excluding slots from basebackup or dropping at start is not enough? See [this Slack thread](https://neondb.slack.com/archives/C07S7RBFVRA/p1729161617307559?thread_ts=1729113766.496999&cid=C07S7RBFVRA)

Either way, it's not a matter of this PR, so I added an item to the main Epic about outgoing LR https://github.com/neondatabase/neon/issues/6213